### PR TITLE
Improve the design doc of Docker build

### PR DESCRIPTION
**Pull request by wangkuiyi**
Fixes https://github.com/PaddlePaddle/Paddle/issues/1625

### CODE REVIEW

File: paddle/scripts/docker/README.md [hunk 1]
```diff
@@ -1,38 +1,146 @@
...
+   This is the production image, generated using the development image. This image might have multiple variants:
+
+   - GPU/AVX   `paddle:<version>-gpu`
+   - GPU/no-AVX  `paddle:<version>-gpu-noavx`
+   - no-GPU/AVX  `paddle:<version>`
+   - no-GPU/no-AVX  `paddle:<version>-noavx`
+
+   We'd like to give users choices of GPU and no-GPU, because the GPU version image is much larger than then the no-GPU version.
+
+   We'd like to give users choices of AVX and no-AVX, because some cloud providers don't provide AVX-enabled VMs.
```
Comment (gangliao) [inline paddle/scripts/docker/README.md]: Runtime check SIMD design docs for containers #1607

File: paddle/scripts/docker/README.md [hunk 2]
```diff
@@ -1,38 +1,146 @@
...
+1. Build a production image
+
+    On the host, we can use the `$PWD/build/Dockerfile` to generate a production image.
+
+   ```bash
+   docker build -t paddle --build-arg "BOOK=ON" -f build/Dockerfile .
    ```
 
-   其中 paddlectl 应该是我们自己写的一个脚本，调用kubectl来在Kubernetes机群上启动一个job的。
+1. Run the Paddle Book
```
Comment (typhoonzero) [inline paddle/scripts/docker/README.md]: The production image better to be smaller, and not all trainning jobs require paddle book to run. May be a good choice to put paddle book into "develop" docker image? For the book may only be useful when developing or learning to develop.

File: paddle/scripts/docker/README.md [hunk 3]
```diff
@@ -1,38 +1,146 @@
...
+
+   Once we have the production image, we can run [Paddle Book](http://book.paddlepaddle.org/) chapters in Jupyter Notebooks (if we chose to build them)
 
+   ```bash
+   docker run -it paddle
```
Comment (typhoonzero) [inline paddle/scripts/docker/README.md]: Paddle production image default command can be `paddle version` which prints out the version and exit. Users can mount a volume contains their real trainning job or build their own docker image based on the production image.

File: paddle/scripts/docker/README.md [hunk 4]
```diff
@@ -1,38 +1,146 @@
...
+   ```
+
+   Note that the default entrypoint of the production image starts Jupyter server, if we chose to build Paddle Book.
+
+1. Run on Kubernetes
+
+   We can push the production image to a DockerHub server, so developers can run distributed training jobs on the Kuberentes cluster:
```
Comment (typhoonzero) [inline paddle/scripts/docker/README.md]: "a DockerHub server" -> DockerHub. DockerHub is know as a service, push to a server may confuse? Not sure about that.

File: paddle/scripts/docker/README.md [hunk 5]
```diff
@@ -1,38 +1,146 @@
...
+1. Build/update the development image (if not yet)
 
    ```bash
-   docker tag paddle yiwang/paddle:did-some-change
-   docker push
-   paddlectl run yiwang/paddle:did-some-change /paddle/demo/mnist/train.py
+   docker build -t paddle:dev . # Suppose that the Dockerfile is in the root source directory.
+   ```
+
+1. Build the source code
```
Comment (typhoonzero) [inline paddle/scripts/docker/README.md]: Should we build paddle when building the develop docker image? So developers will get to work immediately before they make some modifications and rebuild paddle again?

File: paddle/scripts/docker/README.md [hunk 6]
```diff
@@ -1,38 +1,146 @@
...
+   - document writers / Website developers -- Our documents are in the source repo in the form of .md/.rst files and comments in source code.  We need tools to extract the information, typeset, and generate Web pages.
+
+   So the development image must contain not only source code building tools, but also documentation tools:
+
+   - gcc/clang
+   - nvcc
+   - Python
+   - sphinx
+   - woboq
+   - sshd
```
Comment (Yancey1989) [inline paddle/scripts/docker/README.md]: Maybe the `sshd` service is not necessary, `docker exec` can also allow user get into the container with multiple terminals, and in kubernetes cluster,  it is  inconvenient to  allocate a port for PaddlePaddle pod.

File: paddle/scripts/docker/README.md [hunk 7]
```diff
@@ -1,38 +1,146 @@
...
+   - release engineers -- use this to build the official release from certain branch/tag on Github.com.
+   - document writers / Website developers -- Our documents are in the source repo in the form of .md/.rst files and comments in source code.  We need tools to extract the information, typeset, and generate Web pages.
+
+   So the development image must contain not only source code building tools, but also documentation tools:
+
+   - gcc/clang
+   - nvcc
+   - Python
+   - sphinx
+   - woboq
```
Comment (reyoung) [inline paddle/scripts/docker/README.md]: woboq看起来不是必须的开发工具。

File: paddle/scripts/docker/README.md (hunk 4 above)
Comment (Yancey1989) [inline paddle/scripts/docker/README.md]: "a DockerHub server" => Docker Registry such as DockerHub?

File: paddle/scripts/docker/README.md (hunk 2 above)
Comment (helinwang) [inline paddle/scripts/docker/README.md]: We could use `book` tag as well. Feels like book is more for users trying to learn deep learning than developers.

**Comment by reyoung** [comment]
不是很理解为什么我们的production image一定要是用编译时生成的Dockerfile来进一步生成？

production image在我看来应该非常简单。

1. Develop image可以编译出deb安装包。
1. production image只是`FROM ubuntu:14.04`然后`apt install`这个deb安装包即可。不应该再有过多的逻辑。

File: paddle/scripts/docker/README.md (hunk 5 above)
Comment (helinwang) [inline paddle/scripts/docker/README.md]: I think it's a good idea!

**Comment by wangkuiyi** [comment]
@reyoung 因为我连安装 deb 这个包的Dockerfile都不想放在 git repo里。

File: paddle/scripts/docker/README.md (hunk 2 above)
Comment (wangkuiyi) [inline paddle/scripts/docker/README.md]: Good point.

Currently, we have a git submodule in `paddle` repo that points to `book`. Indeed, its that `book` depends on `paddle`. So we should have a git submodule in `book` repo that points to `paddle`.  After we change this, we could have a Dockerfile in `book` repo. This Dockerfile builds Docker image `paddlepaddle/book:<version>` basing on `paddlepaddle/paddle:<version>` built from `paddle` repo. @gangliao

File: paddle/scripts/docker/README.md (hunk 3 above)
Comment (wangkuiyi) [inline paddle/scripts/docker/README.md]: Good point.

I am adding more sections describing how to write a PP program using the development package and how to release it using the production image.

File: paddle/scripts/docker/README.md (hunk 4 above)
Comment (wangkuiyi) [inline paddle/scripts/docker/README.md]: Good point! Rename it to Docker registry.

File: paddle/scripts/docker/README.md (hunk 7 above)
Comment (wangkuiyi) [inline paddle/scripts/docker/README.md]: I'd like to have it when I am facing the currently messy C++ codebase. Maybe we can remove it later after our re-implementation work.

File: paddle/scripts/docker/README.md (hunk 6 above)
Comment (wangkuiyi) [inline paddle/scripts/docker/README.md]: I'd like to have it. I noticed that many people are developing using a Linux box with GPUs.  If we don't have sshd in the container, developers would have to do two steps: (1) ssh to the box, and (2) run docker exec.

File: paddle/scripts/docker/README.md (hunk 5 above)
Comment (wangkuiyi) [inline paddle/scripts/docker/README.md]: OK. Done.

Comment (typhoonzero) [inline paddle/scripts/docker/README.md]: Got it. Since developers has to follow the build instructions at README.md, this will be clear enough for users.

**Pull request merged**